### ISSUE TITLE
v6.x and XE2 etc.

### ISSUE THREAD

**Issue by adem0x** [JAM-Software/Virtual-TreeView#614]
I really do not see why you simply issue a horrible error message wrapped around with a compiler directive that VT 6.x does not support anything less than XE3 when in fact (using a few IFDEFs) it does --I have done it and it works.

Secondly, could you please also make sure that the projects under the Demos folder compile and run. They are always useful as a starting points for learning/remembering VT. Plus, as you _**are**_ distributing them along with the component, it does not make sense to include uncompilable demos.

**Comment by vincentparrett** [comment]
I'm going to assume you do not understand how open source works. The project is run by volunteers, if you don't like it don't use it. You don't get to demand anything. 

If you think you can make it work with XE2, then submit a pull request for consideration. If the demos don't compile, fix them and submit a separate pull request. Your tone is not going to make the maintainers rush to help you.

**Comment by romankassebaum** [comment]
@Vincent: +1!!!!!

**Comment by landrix** [comment]
Incredible.

**Comment by romankassebaum** [comment]
What is incredible?

**Comment by landrix** [comment]
the behavior of the first writer is incredible.

**Comment by romankassebaum** [comment]
I know Joachim privately and he spends a lot of his spare time for this library. I really hold him in regard.

**Comment by adem0x** [comment]
Guys, before you join in the snap-back party, let me remind you that there's an overriding principle: "If it aint broke don't fix it!"

IOW, just because something is open source, you don't get have free reign to break years-long compatibility expecting someone else to fix it back. It is a simple question of discipline.

Same with the Demos: It isn't exactly rocket science to create a program group that also includes the demos such that (on bulding all of it) you can immediately see if that nice and shiny change you've just made broke the demos --and fix them.

**Comment by vincentparrett** [comment]
_sigh_ @adem0x you just don't get it do you. Go away and learn how to be polite and respectful of other people's efforts - http://spf13.com/post/how-to-be-a-good-open-source-community-member/

**Comment by joachimmarder** [comment]
This has been thoroughly discussed in issue #503, there is no point in opening a new issue.

@adem0x :There is a difference in "support" and "somehow getting it to compile". I try to do the first one. I simply can't support a Delphi version that I don't use or have installed.

An XE2 branch has been opened and anyone is welcome to maintain it. So far non of the people who complained about the missing XE2 support volunteered to do this. It seems complaining is much easier.

> Secondly, could you please also make sure that the projects under the Demos folder compile and run.

Believe it or not: I am running a continuous integration server using Jenkins that compiles the packages and demos after each push with some Delphi versions. So I am trying to make sure that they compile. But what is the point in reporting this issue, but not including the Delphi version, the project's name, and the compiler error??

**Comment by adem0x** [comment]
The first thing I do not understand was why you chose to remove things from VT's original codeset when you could have wrapped the new additions with compiler defines.

Another thing I do not understand is why you seem to prefer eye-candies that only exist in the newest versions of Delphi. 

I mean, what is so great about "StaticText.IsEmpty" that "StaticText = ''" couldn't do?

It cannot be the savings on keystrokes --the first one is 18 chars, while the latter is 15.

If this isn't forced obsolescence, what is?..

On the subject of support: Obviously, no one can force you to support older versions of Delphi. 

Except that, it wouldn't have taken much effort to NOT break the already working code and the demos --the Advanced Demo, in particular, which was always both a test for the code as well as a teaching tool.

Assuring compatibility with the older versions of Delphi is not terribly hard: If you have a license to the newest version of Delphi, you do automatically have access/license to older version --I am sure you know that. 

All you would have to do, then, would be to create a couple of VMs with those versions of Delphi installed --D7, XE (or XE2, a better choice, IMO) and probably D2009. 

You would not have to run build farms in them --a quick compile and run to check to see things are OK at the time of a new release time would do. This is what I would have done, and exactly what I do with my projects.

Now, on the suggestions of forking..

Of course, it could be done --and given the state/attitude of VT 6.x development, I wouldn't be surprised if it happened. 

Yet, I wouldn't be pleased if it did, simply because it just means fragmentation of a community that was never big enough to start with.

What am I, personally, going to do about this?

Well.. given the /tone/ of the crowd here, I doubt I will make the effort to do the things I have mentioned (fix the things you deliberately/recklessly broke/omitted) and then come begging that my changes be applied in the trunk here.

BTW, I did read the thread you gave link to. Nothing was concluded because you collectively ignored the 'canaries in the coal mine'.

**Comment by joachimmarder** [comment]
> you could have wrapped the new additions with compiler defines.

Because I can't test if this code compiles and works.

> "StaticText.IsEmpty" that "StaticText = ''" couldn't do?

I like using record helpers becuase they make code mor readable in my opiniton. I use them all the time after I switched to XE3 and I don't want to change my coding style here just becuase I work on Virtual TreeView.

> If this isn't forced obsolescence, what is?..

Getting paranoid? Is it really so hard to believe that a developer likes the concept of record helpers to make code more readable?

> Except that, it wouldn't have taken much effort to NOT break the already working code a

My compiler won't tell if I do or not, that was always the problem. Then people open issues for Delphi versions I don't have. That's simply frustrating. 

> All you would have to do, then, would be to create a couple of VMs with those versions of 
> Delphi installed --D7, XE (or XE2, a better choice, IMO) and probably D2009. 

This is what I did a couple of years and I know that it ate too much of the time I can spend for Virtual TreeView, and of my disk space. If this is so simple, why don't you do it? 

> Well.. given the /tone/ of the crowd here

I don't think the tone in general is bad here if you look into the other issue. But if people come to this issue tracker and say: "Hey, spend your spare time doing this and that so that I don't have to it", then you shouldn't be surprised if the response isn't too friendly.

I have a few hours every week I can spend for VTV and I decided not to spent with maintaining code for old Delphi versions, but to improve the code and the control.

> come begging that my changes be applied in the trunk here.

The concept of branches and forks does not make this necessary. You can pull with the need to push.

**Comment by adem0x** [comment]
> I like using record helpers because they make code more readable in my opinion. I use them all the time after I switched to XE3 and I don't want to change my coding style here just because I work on VirtualTreeView.

How shall I put this: It is a bit of an etiquette thing.. I mean, if the original code was yours, by all means, have your day and let the code reflect your personality and taste. But, it isn't and people who have been maintaining the VT code for all these years have done so in a specific way for a reason.

One more thing I am surprised is this: From between the lines of your reply, it is obvious that you are not such a wet-behind-the-ears Delphi developer; if that's so, it is surprising that you do not --instinctively, so to speak-- know what what would break older Delphi versions.

> This is what I did a couple of years and I know that it ate too much of the time I can spend for Virtual TreeView, and of my disk space. If this is so simple, why don't you do it? 

I think you missed it when I said "exactly what I do with my projects".

OTOH, I have no idea what sort of HW you have access to; so you might have all valid reason. Over here, however, I have movie files that are bigger than VMs.

> I have a few hours every week I can spend for VTV and I decided not to spent with maintaining code for old Delphi versions, but to improve the code and the control.

But, by the look of things, we can hardly speak of " improve the code and the control", can we? What we have here is almost a new creature..

> The concept of branches and forks does not make this necessary. You can pull with the need to push.

I am not sure Delphi is conducive to Git (or any such thing) especially where I am involved: I do have my own code formatting style (no hanging begins, no hanging case elements etc.) and that make diffing almost completely useless. 

As a result, I prefer to report bugs in short snippets explaining where they apply and why.

Anyway, it was good to talk to you. I wish you many happy hours with morphing VT beyond recognition :)

**Comment by joachimmarder** [comment]
> , it is surprising that you do not --instinctively, so to speak-- know what what would break older Delphi versions.

When using generics, anonymous procedures and record helpers I know exactly which Delphi version I break, but I still want to use them when I write code in year 2016. When it comes to more subtle changes in the VCL, e.g regarding VCL styles as seen from XE2 to XE3, I agree: I simply don't have the entire VCL of the past 10 Delphi versions in my mind, and so I unintentionally wrote code that broke with XE or XE2 in the past.

> I think you missed it when I said "exactly what I do with my projects".

Got that, but it does not explain why you don't want to do this for Virtual TreeView? The commits that break with XE2 shouldn't be that many, for someone who is using XE2 anyway, it should cost only a few minutes each month to fix those lines in an XE2 branch.

> I wish you many happy hours with morphing VT beyond recognition

For those who don't want to move forward, the V5 branch is still there and seeking a maintainer. Using this code actively maintained until 2015 with a VCL from 2011 or older does not sound that bad. Nobody is forced to follow my route, but almost all feedback was positive and encouraged me to go further this way.

**Comment by joachimmarder** [comment]
There is meanwhile an XE and XE2 compatible branch maintained by @sglienke at:
https://github.com/sglienke/Virtual-TreeView

**Comment by adem0x** [comment]
Thank you :)

**Comment by adem0x** [comment]
Bear with me one lats time --I am just too much of a n00b as far as Git/Github is concerned: 

I have modified sglienke's codes so that they build/compile (and produce no warnings) (and Demos compile and run) under XE2, but there does not seem to be an 'Issues' tag under sglienke's repo to report these changes.

What would be the best way to go around this?

Ask him/her by email to open/enable 'Issues', or do I _**have to**_ fork? [obviously, I am trying to avoid the latter.]

**Comment by sglienke** [comment]
I am using the VTV under XE (not XE2) at work thus I made sure I have the latest version working there and thought I share it because it might be useful to someone. I don't care for the demos nor giving any support for it. Therefor I am not going to open the issue tracker on the fork.

Personally I am not very happy about the decision to entirely drop support for versions earlier than XE3  (because imo supporting down to XE does not require much effort at all) but so be it.

**Comment by adem0x** [comment]
Thank you for chiming in.

> I don't care for the demos nor giving any support for it. Therefor I am not going to open the issue tracker on the fork.

OK, that's good enough of an answer for me.

I'll fork yours and see --in time-- if I can manage to upload my changes.

**Comment by joachimmarder** [comment]
> because imo supporting down to XE does not require much effort at all

This may be right for a non-UI component but is this is definitely wrong for Virtual Treeview. Getting the code right so that it works well with a VCL with and without styles made so many $ifdefs necessary that this code was really hard to maintain and this ate up too much of my time. XE2 had some V1.0 VCL style bugs, working around them made the code even more difficult to maintain. Sometimes code for XE, other code for XE2 and again other code for XE3+ was needed. It may not be that obvious if you use Virtual TreeView just in 1 or 2 XE projects, but this is different when people use all features of Virtual TreeView in all compilers.

By the way, gave you tried if your code compiles on XE2? I think it didn't support the property "StyleEelments".

In short: It is not much effort to support just XE if you use it for your daily work and I would appreciate if some does this. But it requires quite some effort to support XE - 10 Seattle with a single source code if you do not use XE and XE2 for work.

**Comment by sglienke** [comment]
> have you tried if your code compiles on XE2?

No, I don't have XE2 installed.

> Getting the code right so that it works well with a VCL with and without styles made so many $ifdefs necessary that this code was really hard to maintain.

Since the introduction of VCL styles it was just a matter of coding some helpers for non style versions to make them behave the same as without styles on the versions that support them ime.

**Comment by adem0x** [comment]
After I forked sglienke's repo, I noticed that in the README.md it says that JAM Software (the repo here) is the new project owner. Is this something official --i.e. did Mike Lischke have any say in this.

Another thing about the Demos:

**Comment by adem0x** [comment]
> This may be right for a non-UI component but is this is definitely wrong for Virtual Treeview.

You might want to tell that to the Jedi team --apparently they have been doing all sorts of /wrong/s over all these years.

> Getting the code right so that it works well with a VCL with and without styles made so many $ifdefs necessary that this code was really hard to maintain and this ate up too much of my time.

IOW, instead of spending some time figuring out what each compiler define meant, you chose to strip them all off..

How commendable.

> XE2 had some V1.0 VCL style bugs, working around them made the code even more difficult to maintain. Sometimes code for XE, other code for XE2 and again other code for XE3+ was needed. It may not be that obvious if you use Virtual TreeView just in 1 or 2 XE projects, but this is different when people use all features of Virtual TreeView in all compilers.

This is precisely what compiler directives are for. You put all the relevant ones in an .inc file and use them in the source. For inspiration, you could look at, for example, FastMM4's inc file. Or those of the Jedi project.

> By the way, have you tried if your code compiles on XE2? I think it didn't support the property "StyleEelments".

It didn't at first. Adjusting the compiler directives in a few places sorted that out easily.

> In short: It is not much effort to support just XE if you use it for your daily work and I would appreciate if some does this. But it requires quite some effort to support XE - 10 Seattle with a single source code if you do not use XE and XE2 for work.

I don't agree. It probably was more effort to remove all the existing conditional directives etc. than adding a couple. You could then add your newer features much more easily without breaking all backwards compatibility.

Finally, what is the point of including uncompilable code in the repo? Why don't you remove all the Demos?

One more thing: Since you don't care about them; in the name of not distributing uncompilable code, shouldn't you remove the demos all together.

You might also decide to not call Jam Software 'the new project owner' since you disown all of its past.

**Comment by joachimmarder** [comment]
>  noticed that in the README.md it says that JAM Software (the repo here) 
> is the new project owner. Is this something official --i.e. did Mike Lischke have any say in this.

Yes, of course this was official, as a quick view on his homepage would have told you:
http://www.soft-gems.net/index.php/controls/virtual-treeview

**Comment by joachimmarder** [comment]
> You might want to tell that to the Jedi team --apparently they 
> have been doing all sorts of /wrong/s over all these years.

Please read my post before you answer: I did not write that it is wrong to support old versions, I just wrote that it is not as easy for UI controls as it is for RTL code, especially for custom drawn controls if style and non-style Delphi versions should be supported in one code.

I appreciate if someone does it. If you think there should be an XE2 version of Virtual TreeView, then simply do it instead of wasting time with such posts and tell other people, they should do what you want in their spare-time instead of playing with their kids.

>  instead of spending some time figuring out what each compiler define meant, you chose to strip them all off..

Exactly! Spending time to move the project forward instead of wasting time with $ifdefs for old compiler versions I don't have, this is what makes sense.

> Finally, what is the point of including uncompilable code in the repo?

What code is uncompatible with what exaclty? Why don't you simply open an issue instead o making such vague statements?

> You might also decide to not call Jam Software 'the new project owner' since you disown all of its past.

Great, this is really motiviating after I invested hundreds of hours of my time in V5, which still is there and works fine with XE and XE2.

**Comment by joachimmarder** [comment]
> No, I don't have XE2 installed.

Interesting. But you think I should have? Or should I support XE2 without having it installed?

**Comment by sglienke** [comment]
> Interesting. But you think I should have? Or should I support XE2 without having it installed?

You are barking up the wrong tree - I never asked you to support anything. I only commented on some error prone record helper usage. You were the one that brought up my fork in this thread and assumed it supports XE2 (which I never said) and I commented on questions.

**Comment by joachimmarder** [comment]
> I never asked you to support anything

Hmm, this sounded at least very closely:

> Personally I am not very happy about the decision to entirely
> drop support for versions earlier than XE3

That was what I was referring to, not your comment about record helper, to which I agree in some part.

**Comment by the-Arioch** [comment]
> If you think there should be an XE2 version of Virtual TreeView, then simply do it

Well, at least you do advertise Stefan's fork as "XE2-compatible" on VTV page  and it also comes close to promise of XE2 support on his part.

**Comment by joachimmarder** [comment]
> Well, at least you do advertise Stefan's fork as "XE2-compatible" on VTV page

Sure, why should I have a problem with problem with an XE2 fork? I just don't want to do XE2 and below support myself.

**Comment by the-Arioch** [comment]
AFAIR you did when I uploaded my XE2 fork and told you about both my and Stefan's repositories. So perhaps I am kind of jealous. But that would be nothing... What bugs me here is 

> You were the one that brought up my fork in this thread and assumed it supports XE2 (which I never said)

See...  You do not want to be the one considered responsible for XE2 which you do not use nor do want to use. But so does Stefan.

You offloaded repeating XE2 noise to him because you do not want to deal with it time and again. And I thought he likes more exposure to his project. But now I wonder if he would prefer to be mentioned only within XE1 context.

That said, he could edit readme.md in his repo and remove XE2 mention, but he did not, so perhaps he de facto agreed with being collector of both XE1 and XE2 users :-)

**Comment by the-Arioch** [comment]
> Or those of the Jedi project.

Dear adem0x, did you ever wondered when and why JediVCL project stopped support for D5 and BCB5 ?

You brought a good point but if you would learn a bit how that happened for the first time and how that happened finally - then you would have to agree, that JEDI precedent you brought supports Vincent and Joachim not you.

**Comment by joachimmarder** [comment]
> You do not want to be the one considered responsible for XE2 which you do not use.

Exactly

> So perhaps I am kind of jealous

If you like to be mentioned with your fork, I don't have any problem with that and will be happy to do so. I can link Stefan' s fork for XE and yours for XE2 if you like.

**Comment by the-Arioch** [comment]
Let's see if Stefan maybe says anything.

To the moment I developed mixing feelings.

On one hand in near future I would hardly abandon XE2 and at least would keep VTV running for my own needs, like I did with JVCL for D5 back then.

OTOH since you marked Stefan's fork as faithful backport, I started taking some liberty in deviating from upstream, and we do not always agree over their merits. So my fork balances on becoming experimental rather than stable branch and ATM I ain't sure if it would increase or fade.

Frankly, if I'd found Stefan's port before forking of my own I'd most probably would not even bother :-)

**Comment by Fr0sT-Brutal** [comment]
@adem0x, your point is like was mine in #503 :) but authors are the gods, and they have the right to drop support of older versions. F.i., check out the maintained versions of old TurboPack projects - they support only one recent compiler!
@joachimmarder, could you add a link to my repo to Readme? It surely supports XE2 and implements this support in the form of several commits that could be easily merged into base branch.
@All, welcome to my fork https://github.com/Fr0sT-Brutal/VirtualTreeView_mod/tree/fr0st_xe2. I've recently rebased it to the newest changes.

**Comment by sglienke** [comment]
<del>FWIW I am using my forks version for TestInsight and it also compiles and works (granted I don't use any fancy things or VCL styles) on XE2.<del>

Edit: forget that I was stupid - I use the v5_stable there

Personally I am always in favor of sticking to the original repo but I respect when people don't want into the nightmare of supporting too many different versions that they don't even have installed anymore (I myself don't have XE2-4 installed for example and always have to run tests for code that supports these versions on some different machine before release and that is non VCL code so easily testable by unit tests).
However when I need a fork I always aim at being as less intrusive as possible to also make it easier for me to keep up and rebase on never origins.
Also if I add new features I make sure that I have a clean cut between the "support version X" code and my added things.

**Comment by joachimmarder** [comment]
> @joachimmarder, could you add a link to my repo to Readme? 

Done.

I still think having a fork for a single old Delphi version maintained by people that actively use this Delphi version will result in better quality compared to a situation in which a single developer needs to maintain a single source UI control for 10 (soon more) Delphi versions.

**Comment by Fr0sT-Brutal** [comment]
> I still think having a fork for a single old Delphi version maintained by people that actively use this Delphi version will result in better quality compared to a situation in which a single developer needs to maintain a single source UI control for 10 (soon more) Delphi versions.

I tend to agree though this could be sad for people stuck on older versions. Using new cool features is certainly much more fun than supporting ancient stuff.

**Comment by joachimmarder** [comment]
>this could be sad for people stuck on older versions

But here the root problem seems to be that they are stuck with an old Delphi version, not that new features are not added to old versions (which still could be done by volunteers using that old Delphi version).

**Comment by Fr0sT-Brutal** [comment]
Joachim, you're right, backporting is a widely used thing. Bad times for Delphi and later Emba's policy forced some people to stay on older versions (I haven't heard anyone is stuck on ancient FPC version) but they usually can support older compilers themselves (except if the code is changed significantly).

**Comment by the-Arioch** [comment]
I caught up my XE2 repo. Had one single conflict (VTV version string constant) with changes from 04-MAY-2016 till 18-MAR-2017. Either because VTV code quality, or due to non-XE1 approach I had chosen from starters, so all the record-helpers methods are source-compatible with me and need not be replaced everywhere by global routines.

Made two small fixes also: fb77e8a16599eb73944ce131fc9c6cbc8151b967 and next c74c0328be079e764d11eee130907c545047bc61

**Comment by the-Arioch** [comment]
Well, XE2 is the "devil I know" https://github.com/the-Arioch/XE2fixes

Upgrading Delphi - we have an XE4 license in office, but use it to run another XE2 instead - means I should be dropped with covered eyes into the ocean of new bugs. Maybe there are very few of them in some particular "lucky" release. Or maybe not. It is a game of chance.

It is not just the price policy that matters (while price too), it is the outcome - I would have to invest unpredictable time into moving the codebase to new Delphi and then ironing suddenly popping RTL/VCL bugs here and there. And if I would finally decide to give up on it - how would I backport my old codebase (including nested inherited DFMs) to older Delphi ?

**Comment by joachimmarder** [comment]
> Made two small fixes also: fb77e8a and next c74c032

Can you make a pull request?

>or due to non-XE1 approach I had chosen from starters

That certainly helps a lot! XE2 and XE3 are quiet code compatible (compared to XE1). But XE2 has some nasty bugs in VCL themes and the 64Bit compiler. In case you don't need both, then XE2 is fine.

**Comment by the-Arioch** [comment]
> Can you make a pull request?

Hardly so: that request would pull ALL the commits i made after forked from you year ago. Every of them.
So, if you want to pul from my repo, just to discard most of my changes, for the mere sake of code trees thinking they were solicited and synced - you can try it. But no, i never managed to discipline myself so well, that my post-forking changes would be isolated.

Okay, i could make in my repo the "XE3 branch" - your vanilla one re-forked - and try backporting my XE2 chanegs to it for the sake of limited isolated pull requests - but having no XE2 i would not even be able to compile it.

>> or due to non-XE1 approach I had chosen from starters

> That certainly helps a lot!

Maybe, or maybe not. That would take downloading Frost's repo and comparing to your one to see how many actual changes were needed.
Quite possible, after his (or rather Stefan's) initial reverting helpers' methods back to global functions, that after that initial burst of editings, the process of catching up was as simple as mine.
